src/interface/swarm_device_context.hpp

```diff
@@ -0,0 +1,108 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+#define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+
+#include "coordinates/coordinates.hpp"
+
+namespace parthenon {
+
+class SwarmDeviceContext {
+ public:
+  KOKKOS_FUNCTION
+  bool IsActive(int n) const { return mask_(n); }
+
+  KOKKOS_FUNCTION
+  bool IsOnCurrentMeshBlock(int n) const { return blockIndex_(n) == this_block_; }
+
+  KOKKOS_FUNCTION
+  void MarkParticleForRemoval(int n) const { marked_for_removal_(n) = true; }
+
+  KOKKOS_FUNCTION
+  bool IsMarkedForRemoval(const int n) const { return marked_for_removal_(n); }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetNeighborBlockIndex(const int &n, const double &x, const double &y,
+                            const double &z, bool &is_on_current_mesh_block) const {
+    int i = static_cast<int>(std::floor((x - x_min_) / ((x_max_ - x_min_) / 2.))) + 1;
+    int j = static_cast<int>(std::floor((y - y_min_) / ((y_max_ - y_min_) / 2.))) + 1;
+    int k = static_cast<int>(std::floor((z - z_min_) / ((z_max_ - z_min_) / 2.))) + 1;
+
+    // Something went wrong
+    if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
+        ((k < 0 || k > 3) && ndim_ > 2)) {
+      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+    }
+
+    // Ignore k,j indices as necessary based on problem dimension
+    if (ndim_ == 1) {
+      blockIndex_(n) = neighborIndices_(0, 0, i);
+    } else if (ndim_ == 2) {
+      blockIndex_(n) = neighborIndices_(0, j, i);
+    } else {
+      blockIndex_(n) = neighborIndices_(k, j, i);
+    }
+
+    is_on_current_mesh_block = (blockIndex_(n) == this_block_);
+
+    return blockIndex_(n);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetMyRank() const { return my_rank_; }
+
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &x, const Real &y, const Real &z, int &i, int &j, int &k) const {
+    i = static_cast<int>(
+            std::floor((x - x_min_) / coords_.Dx(CoordinateDirection::X1DIR))) +
```

Comment:
This works for uniform cartesian.
I suggest (again not blocking), to use `coords_.dx1v(i)` (and similarly `coords_dx2v(j)` and `coords_dx3v(k)`) as its the more common usage (at the same computational cost).

Reply:
I'm not sure we can do the usual `coords_.dx1v(i)` here, since `i` is what we're calculating. Technically it would work I think because `i`, `j`, `k` aren't actually used in `uniform_cartesian.hpp`'s `dx*v()` functions I don't think, but it could be quite confusing. Since the behavior is a bit different I slightly prefer keeping the unusual `Dx()` call here.

Thinking more on this `Xtoijk` in the general case would have to become a nasty binary search over the dx's of each zone in each direction to figure out what your `i` actually is... an alternative would be to have each particle keep a record of the zone index its currently in, and require particle push functions to update that appropriately. What this actually makes me think is that particle support for non-uniform meshes will be even further out in the future.

Reply:
Not now, but maybe for the future, `Xtoijk` should be part of the `Coords` class. Then the implementation details can be punted to the coordinate system.

Reply:
Yeah good idea

Reply:
My bad. Using `dx1v` defeats the purpose here. I agree that it'd be nice to have this functionality in `Coords` at some point.